Mathlib/Data/Nat/Find.lean

```diff
@@ -24,29 +24,27 @@ private def lbp (m n : ℕ) : Prop :=
 
 variable [DecidablePred p] (H : ∃ n, p n)
 
-private def wf_lbp : WellFounded (@lbp p) :=
-  ⟨let ⟨n, pn⟩ := H
-    suffices ∀ m k, n ≤ k + m → Acc lbp k from fun a => this _ _ (Nat.le_add_left _ _)
-    fun m =>
-    Nat.recOn m
-      (fun k kn =>
-        ⟨_, fun y r =>
-          match y, r with
-          | _, ⟨rfl, a⟩ => absurd pn (a _ kn)⟩)
-      fun m IH k kn =>
-      ⟨_, fun y r =>
-        match y, r with
-        | _, ⟨rfl, _a⟩ => IH _ (by rw [Nat.add_right_comm]; exact kn)⟩⟩
-
 protected def findX : { n // p n ∧ ∀ m < n, ¬p m } :=
-  @WellFounded.fix _ (fun k => (∀ n < k, ¬p n) → { n // p n ∧ ∀ m < n, ¬p m }) lbp (wf_lbp H)
-    (fun m IH al =>
-      if pm : p m then ⟨m, pm, al⟩
-      else
-        have : ∀ n ≤ m, ¬p n := fun n h =>
-          Or.elim (Nat.lt_or_eq_of_le h) (al n) fun e => by rw [e]; exact pm
-        IH _ ⟨rfl, this⟩ fun n h => this n <| Nat.le_of_succ_le_succ h)
-    0 fun n h => absurd h (Nat.not_lt_zero _)
+  go H 0 (by simp)
+where
+  go (H : ∃ n, p n) (n : ℕ) (H' : ∀ n' < n, ¬ p n') : { n // p n ∧ ∀ m < n, ¬p m } :=
+    if h : p n then ⟨n, h, H'⟩
+    else
+      go H (n + 1) <| by
+        intro n' hn'
+        obtain hn' | rfl := lt_or_eq_of_le (le_of_lt_succ hn')
+        · exact H' n' hn'
+        · exact h
+  termination_by H.choose - n
```

Comment:
This adds a dependency on the axiom of choice, which is perhaps unfortunate and something @digama0  has opinions about.